lib/gc/Transforms/MergeNestedForall.cpp

```diff
@@ -0,0 +1,94 @@
+//===-- MergeNestedForall.cpp - Merge nested scf.forall op ------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace mlir {
+namespace gc {
+#define GEN_PASS_DEF_MERGENESTEDFORALL
+#include "gc/Transforms/Passes.h.inc"
+
+namespace {
+
+struct MergeNestedForallLoops : public OpRewritePattern<scf::ForallOp> {
+  using OpRewritePattern<scf::ForallOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::ForallOp op,
+                                PatternRewriter &rewriter) const override {
+    Block &outerBody = *op.getBody();
+    if (!llvm::hasSingleElement(outerBody.without_terminator()))
+      return failure();
+
+    scf::ForallOp innerOp = dyn_cast<scf::ForallOp>(outerBody.front());
```

Comment:
Seems it only supports two level `forall` merge?

Reply:
This is a rewrite pattern and will merge recursively merge any two neighbor forall op if possible. For example
```
forall1
forall2
forall3
---------->
forall1+2
forall3
---------->
forall1+2+3
```

Reply:
I see, thanks.